subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/JarTransformForAgent.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.classpath.transforms;
+
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.internal.classanalysis.AsmConstants;
+import org.gradle.internal.classloader.TransformReplacer.MarkerResource;
+import org.gradle.internal.classpath.ClasspathBuilder;
+import org.gradle.internal.classpath.ClasspathEntryVisitor;
+import org.gradle.internal.classpath.ClasspathWalker;
+import org.gradle.internal.classpath.types.InstrumentingTypeRegistry;
+import org.gradle.util.internal.JarUtil;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.util.jar.Attributes;
+import java.util.jar.Manifest;
+
+import static org.gradle.internal.classpath.transforms.MrJarUtils.isSupportedVersion;
+
+/**
+ * Transformation for agent-based instrumentation.
+ */
+public class JarTransformForAgent extends BaseJarTransform {
```

Comment:
:nail_care: can the `JarTransformForAgent` class be package-private?

Reply:
It can be, fixed